### PR TITLE
feat(content-layer): support references in content layer

### PR DESCRIPTION
**Pull request by ascorbic**
## Changes

Adds support for the `reference()` in content layer schemas. This does unfortunately mean that we can't error out at validation time if the colleciotn name doesn't exist. This is because the content layer loads collections in parallel, and a colleciotn may not have been created at the time the other is syncing. Instead, what we do is introduce a check at runtime, which logs a warning if the colleciton doesn't exist.

## Testing

Adds tests

### CODE REVIEW

File: packages/astro/src/content/runtime.ts [hunk 1]
```diff
@@ -277,7 +294,9 @@ export function createGetEntry({
 		if (store.hasCollection(collection)) {
 			const entry = store.get<DataEntry>(collection, lookupId);
 			if (!entry) {
-				throw new Error(`Entry ${collection} → ${lookupId} was not found.`);
+				// eslint-disable-next-line no-console
+				console.warn(`Entry ${collection} → ${lookupId} was not found.`);
```
Comment (ascorbic) [inline packages/astro/src/content/runtime.ts:298]: There seems to be inconsistency in different parts of the existing code about what the correct behaious should be when `getEntry` is called and there's no match. Should it throw or just warn? I'm inclined to suggest that it throw if there's no matching collection, but just return undefined if there's no matching entry.

File: packages/astro/src/content/runtime.ts [hunk 2]
```diff
@@ -208,6 +218,11 @@ export function createGetDataEntryById({ getEntryImport }: { getEntryImport: Get
 				collection,
 			};
 		}
+		if (!collectionNames.has(collection)) {
+			// eslint-disable-next-line no-console
+			console.warn(`The collection ${JSON.stringify(collection)} does not exist.`);
```
Comment (ascorbic) [inline packages/astro/src/content/runtime.ts:223]: Is this the right behaviour? It's not clear whether a missing collection should throw or just warn.

**Comment by ascorbic** [comment]
!preview contentlayer

File: packages/astro/src/content/runtime.ts (hunk 2 above)
Comment (bluwy) [reply packages/astro/src/content/runtime.ts:223]: It also warns in `main` now so I think warning is fine https://github.com/withastro/astro/blob/026e8baf3323e99f96530999fd32a0a9b305854d/packages/astro/src/content/runtime.ts#L64-L72

Comment (ascorbic) [reply packages/astro/src/content/runtime.ts:223]: Yeah, that's the only reason I used it. I was trying to follow the current behaviour.

File: packages/astro/src/content/runtime.ts [hunk 3]
```diff
@@ -168,6 +170,11 @@ export function createGetEntryBySlug({
 				render: () => renderEntry(entry),
 			};
 		}
+		if (!collectionNames.has(collection)) {
+			// eslint-disable-next-line no-console
+			console.warn(`The collection ${JSON.stringify(collection)} does not exist.`);
```
[the remark this answers is not in the document]
Comment (ascorbic) [reply packages/astro/src/content/runtime.ts:175]: Do you have a suggestion of how best to do that? I can't find any other usage of the AstroLogger in the content runtime, so I'm not sure how best to get one in scope

[the remark this answers is not in the document]
Comment (matthewp) [reply packages/astro/src/content/runtime.ts:175]: Yeah anything inside of runtime code is difficult to have a logger for.

**Pull request merged**